### DISCUSSION TITLE
node项目执行node test.mjs报错TypeError: PAGModule.VectorString is not a constructor

### DISCUSSION THREAD

**Discussion by guileryu** [Tencent/libpag#2103] [Q&A] [answered]
![image](https://github.com/Tencent/libpag/assets/59436015/5a03423c-39ce-46f8-8ffb-0808c266413b)

TypeError: PAGModule.VectorString is not a constructor 怎么解决

**Comment by zhalice2011** [comment, marked as answer]
你好, 这是因为WebAssembly 模块没有正确初始化, 报错提示没有找到 wasm 文件, 请按照文档方法配置, 另外 libpag 使用了不浏览器特有的API 所以不止支持 nodejs 环境.

文档:
调用 libpag.js 上的 `PAGInit()` 方法时，默认会加载 libpag.js 同一目录下的 libpag.wasm 文件。如果你希望把 libpag.wasm 放在其他目录下，则可以使用 `locateFile` 将 libpag.wasm 的路径返回给 `PAGInit()` 方法。如下

```js
const PAG = await window.libpag.PAGInit({
  locateFile: () => {
    if (location.host === 'dev.pag.art') {
      // development environment
      return 'https://dev.pag.art/file/libpag.wasm';
    } else {
      // production environment
      return 'https://pag.art/file/libpag.wasm';
    }
  },
});
```

### EsModule

```bash
$ npm i libpag
```

```js
import { PAGInit } from 'libpag';

PAGInit().then((PAG) => {
  // 实例化 PAG
});
```

**使用 ESModule 引入的方式需要注意，像 Webpack 和 Rollup 等打包工具是默认没有打包 .wasm 文件的。也就是说如果你的项目时 Vue / React 这类使用脚手架构建的项目需要把 node_modules 下的 libpag/lib/libpag.wasm 文件打包到最终产物中，并且使用 `locateFile` 钩子返回 libpag.wasm 文件的路径，这样才能确保在网络请求中能加载到 libpag.wasm 文件**

npm package 中提供了多种构建产物，可以阅读 [这里](./doc/develop-install.md) 了解不同目录下产物的差别。

Demo 项目提 [pag-web](https://github.com/libpag/pag-web) 供了简单的接入示例和 Vue / React / PixiJS 等配置示例， 可以点击 [这里](https://github.com/libpag/pag-web) 查看。

更多的 API 接口可以阅读 [API 文档](https://pag.art/api.html#/apis/web/)。
